crypto/crypto/src/noise.rs

```diff
@@ -0,0 +1,587 @@
+    pub fn new(private_key: x25519::StaticSecret) -> Self {
+        let public_key = x25519::PublicKey::from(&private_key);
+        Self {
+            private_key,
+            public_key: public_key.as_bytes().to_owned(),
+        }
+    }
+
+    /// An initiator can use this function to initiate a handshake with a known responder.
+    pub fn initiate_connection(
+        &self,
+        rng: &mut (impl rand::RngCore + rand::CryptoRng),
+        prologue: &[u8],
+        remote_public: &x25519::PublicKey,
+        payload: Option<&[u8]>,
+    ) -> Result<(InitiatorHandshakeState, Vec<u8>), NoiseError> {
+        // checks
+        if let Some(payload) = payload {
+            if payload.len()
+                > MAX_SIZE_NOISE_MSG - 2 * X25519_PUBLIC_KEY_LENGTH - 2 * AES_GCM_TAGLEN
+            {
+                return Err(NoiseError::PayloadTooLarge);
+            }
+        }
+
+        // initialize
+        let mut h = PROTOCOL_NAME.clone();
+        let mut ck = PROTOCOL_NAME.clone();
+        mix_hash(&mut h, &prologue);
+        mix_hash(&mut h, remote_public.as_bytes());
+
+        // -> e
+        let e = if cfg!(test) {
+            let mut ephemeral_private = [0u8; X25519_PUBLIC_KEY_LENGTH];
+            rng.fill_bytes(&mut ephemeral_private);
+            x25519::StaticSecret::from(ephemeral_private)
+        } else {
+            x25519::StaticSecret::new(rng)
+        };
+        let e_pub = x25519::PublicKey::from(&e);
+        mix_hash(&mut h, e_pub.as_bytes());
+        let mut msg = e_pub.as_bytes().to_vec();
+
+        // -> es
+        let dh_output = e.diffie_hellman(remote_public);
+        let k = mix_key(&mut ck, dh_output.as_bytes())?;
+
+        // -> s
+        #[cfg(test)]
+        let aead = Aes256Gcm::new(GenericArray::clone_from_slice(&k));
+        #[cfg(not(test))]
+        let aead = Aes128Gcm::new(GenericArray::clone_from_slice(&k));
+
+        let msg_and_ad = Payload {
+            msg: &self.public_key,
+            aad: &h,
+        };
+        let nonce = GenericArray::from_slice(&[0u8; 12]);
+        let encrypted_static = aead.encrypt(nonce, msg_and_ad).unwrap(); // this API cannot fail
+        mix_hash(&mut h, &encrypted_static);
+        msg.extend_from_slice(&encrypted_static);
+
+        // -> ss
+        let dh_output = self.private_key.diffie_hellman(remote_public);
+        let k = mix_key(&mut ck, dh_output.as_bytes())?;
+
+        // -> payload
+        #[cfg(test)]
+        let aead = Aes256Gcm::new(GenericArray::clone_from_slice(&k));
+        #[cfg(not(test))]
+        let aead = Aes128Gcm::new(GenericArray::clone_from_slice(&k));
+
+        let msg_and_ad = Payload {
+            msg: payload.unwrap_or_else(|| &[]),
+            aad: &h,
+        };
+        let nonce = GenericArray::from_slice(&[0u8; 12]);
+        let encrypted_payload = aead
+            .encrypt(nonce, msg_and_ad)
+            .map_err(|_| NoiseError::Encrypt)?;
```

Comment:
either unwrap here or propagate error above

Reply:
not sure what you mean here

Reply:
L235:

```rust
let encrypted_static = aead.encrypt(nonce, msg_and_ad).unwrap(); // this API cannot fail
```

L251:

```rust
let encrypted_payload = aead
    .encrypt(nonce, msg_and_ad)
    .map_err(|_| NoiseError::Encrypt)?;
```

why panic on one and propagate on the other? why would the first be infallible and the second fallible? if there is a reason, then a comment explaining would be nice. otherwise, the second looks like it would be infallible as well?